### PR TITLE
Interval miscalculation due to operation not being applied in certain scenarios

### PR DESCRIPTION
**Pull request by marregui**
Fixes https://github.com/questdb/questdb/issues/2006

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 11 / 12 (91.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/PeriodType.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/NowFunctionFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModelBuilder.java|1|1|100.00%|

**Comment by bluestreak01** [comment]
is the fix that `now()` wasn't initialized ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 11 / 12 (91.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/PeriodType.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/NowFunctionFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModelBuilder.java|1|1|100.00%|

**Comment by marregui** [comment]
> is the fix that `now()` wasn't initialized ?

No, it is not a fix after all as I break this description https://questdb.io/docs/reference/function/date-time/#now and now would return a different result, just like systimestamp does.

The problem is that the where clause has 3 and-separated comparisons and unfortunately their order gives different results. Ranges are calculated correctly when now is a constant instead of a runtime constant, however I break the database this way.

Now was initialised in RuntimeIntervalModel.addEvaluateDynamicIntervals line 123 (dynamicFunction.init(null, sqlContext);).

In my solution, there are as many instances of now as now expressions are found in the statement, and each is initialised with some skew with respect to the others.

File: core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java [hunk 1]
```diff
@@ -196,7 +196,7 @@ private void addEvaluateDynamicIntervals(LongList outIntervals, SqlExecutionCont
                     case IntervalOperation.INTERSECT:
                     case IntervalOperation.INTERSECT_BETWEEN:
                     case IntervalOperation.INTERSECT_INTERVALS:
-                    case IntervalOperation.SUBTRACT_INTERVALS:
```
[the remark this answers is not in the document]
Comment (marregui) [reply core/src/main/java/io/questdb/griffin/model/RuntimeIntervalModel.java:199]: apologies

File: core/src/main/java/io/questdb/griffin/engine/functions/date/NowFunctionFactory.java [hunk 2]
```diff
@@ -48,25 +47,24 @@ public boolean isRuntimeConstant() {
...
-        public void init(SymbolTableSource symbolTableSource, SqlExecutionContext executionContext) {
-            executionContext.initNow();
-            context = executionContext;
+        public long getTimestamp(Record rec) {
+            return context.getNow();
         }
 
         @Override
-        public boolean isRuntimeConstant() {
+        public boolean isConstant() {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/date/NowFunctionFactory.java:67]: Concept of `Runtime Constant` was invented specifically for `now()` function. Very surprised to see it's reversed

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/PeriodType.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|3|3|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/date/NowFunctionFactory.java (hunk 2 above)
Comment (marregui) [reply core/src/main/java/io/questdb/griffin/engine/functions/date/NowFunctionFactory.java]: yes, indeed, I have reverted this.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 8 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/PeriodType.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/RuntimeIntervalModel.java|3|3|100.00%|

**Review by ideoma — APPROVED** [review]

**Pull request merged**